src/context/walletConnectContext.tsx

```diff
@@ -0,0 +1,329 @@
+  DEFAULT_PROJECT_ID,
+  DEFAULT_RELAY_URL,
+  SUPPORTED_CHAINS,
+  DEFAULT_METHODS,
+  DEFAULT_EVENTS,
+} from "../constants"
+import ConnexService from "../service/ConnexService"
+import { ActionType, useWallet } from "./walletContext"
+
+/**
+ * Types
+ */
+interface IContext {
+  client: Client | undefined
+  session: SessionTypes.Struct | undefined
+  connect: (
+    network: Network,
+    onSuccess: (session: SessionTypes.Struct) => Promise<void>,
+    onError?: (err: unknown) => void
+  ) => Promise<void>
+  disconnect: () => Promise<void>
+  isInitializing: boolean
+  identifyUser: (
+    network: Network,
+    session: SessionTypes.Struct
+  ) => Promise<Certificate>
+}
+
+/**
+ * Context
+ */
+export const WalletConnectContext = createContext<IContext>({} as IContext)
+
+/**
+ * Web3Modal Config
+ */
+
+const web3Modal = new Web3Modal({
+  walletConnectVersion: 2,
+  projectId: DEFAULT_PROJECT_ID,
+  standaloneChains: SUPPORTED_CHAINS,
+  themeVariables: {
+    "--w3m-z-index": "99999999",
+  },
+})
+
+interface IWalletConnectProvider {
+  children: React.ReactNode
+}
+
+export const WalletConnectProvider = ({ children }: IWalletConnectProvider) => {
+  const [client, setClient] = useState<Client>()
+  const [session, setSession] = useState<SessionTypes.Struct>()
+
+  const [isInitializing, setIsInitializing] = useState(false)
+  const { dispatch } = useWallet()
+
+  const reset = () => {
+    console.log("Resetting WalletConnect state")
+    setSession(undefined)
+
+    ConnexService.clear()
+    dispatch({ type: ActionType.CLEAR })
+  }
+
+  const connect = useCallback(
+    async (
+      network: Network,
+      onSuccess: (session: SessionTypes.Struct) => Promise<void>,
+      onError?: (err: unknown) => void
+    ) => {
+      if (typeof client === "undefined") {
+        throw new Error("WalletConnect is not initialized")
+      }
+
+      let session: SessionTypes.Struct | undefined
+      try {
+        const requiredNamespaces = {
+          vechain: {
+            methods: Object.values(DEFAULT_METHODS),
+            chains: [`vechain:${network}`],
+            events: Object.values(DEFAULT_EVENTS),
+          },
+        }
+
+        const { uri, approval } = await client.connect({
+          requiredNamespaces,
+        })
+
+        if (uri) {
+          // Create a flat array of all requested chains across namespaces.
+          const standaloneChains = Object.values(requiredNamespaces)
+            .map((namespace) => namespace.chains)
+            .flat() as string[]
+
+          web3Modal.openModal({ uri, standaloneChains })
+        }
+
+        //TODO: if user closes modal before approving on mobile state is not resetted
+
+        session = await approval()
+        console.log("Established session:", session)
+
+        setSession(session)
+
+        web3Modal.closeModal()
+
+        await onSuccess(session)
+      } catch (e) {
+        web3Modal.closeModal()
+        onError?.(e)
+
+        throw e
```

Comment:
I think we can remove this. We are already passing the error to the caller with the line above